### PR TITLE
[RIP 30] Support Compaction topic

### PR DESCRIPTION
**Pull request by ltamber**
refer to #3799
the implement was base on 5.0.0-beta

### CODE REVIEW

File: store/src/main/java/org/apache/rocketmq/store/kv/CompactionService.java [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+
+    public GetMessageResult getMessage(final String group, final String topic, final int queueId,
+        final long offset, final int maxMsgNums, final int maxTotalMsgSize) {
+        return compactionStore.getMessage(group, topic, queueId, offset, maxMsgNums, maxTotalMsgSize);
+    }
+
+    @Override
+    public String getServiceName() {
+        return CompactionService.class.getSimpleName();
+    }
```
Comment (RongtongJin) [inline store/src/main/java/org/apache/rocketmq/store/kv/CompactionService.java:77]: 这里getServiceName可能要照着其他store层的ServiceThread写下，通过设置线程名来区分不同broker线程，线程名前缀必须是#BrokerClusterName_BrokerName_BrokerId#，否则BrokerContainer下的broker日志将不会分离。

参考https://github.com/apache/rocketmq/blob/5.0.0-beta/docs/cn/BrokerContainer.md

File: store/src/main/java/org/apache/rocketmq/store/kv/CompactionPositionMgr.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+import org.apache.rocketmq.common.ConfigManager;
+import org.apache.rocketmq.remoting.protocol.RemotingSerializable;
+
+import java.io.File;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class CompactionPositionMgr extends ConfigManager {
+
+    public static final String CHECKPOINT_FILE = "position-checkpoint";
```
Comment (RongtongJin) [inline store/src/main/java/org/apache/rocketmq/store/kv/CompactionPositionMgr.java:27]: 感觉叫checkpoint又是继承configManager有点奇怪，另外position-checkpoint好像不能体现出和compaction相关

**Review by RongtongJin — COMMENTED** [review]
It would be better to add an integration test

File: store/src/main/java/org/apache/rocketmq/store/kv/CompactionPositionMgr.java (hunk 2 above)
Comment (ltamber) [reply store/src/main/java/org/apache/rocketmq/store/kv/CompactionPositionMgr.java:27]: 1. 这里主要是复用configManager持久化的能力
2. 从目录结构来看，position-checkpoint文件路径是在 store/compaction/position-checkpoint，与compaction还是有关系的

File: store/src/main/java/org/apache/rocketmq/store/kv/CompactionService.java (hunk 1 above)
Comment (ltamber) [reply store/src/main/java/org/apache/rocketmq/store/kv/CompactionService.java]: fixed

File: store/src/main/java/org/apache/rocketmq/store/kv/CompactionStore.java [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+                        }
+                        try {
+                            String topic = fileTopic.getName();
+                            int queueId = Integer.parseInt(fileQueueId.getName());
+
+                            if (Files.isDirectory(Paths.get(compactionCqPath, topic, String.valueOf(queueId)))) {
+                                CompactionLog log = new CompactionLog(defaultMessageStore, topic, queueId,
+                                    offsetMapSize, positionMgr, compactionLogPath, compactionCqPath);
+                                compactionLogTable.put(topic + "_" + queueId, log);
+                                compactionSchedule.scheduleWithFixedDelay(log::doCompaction, compactionInterval, compactionInterval, TimeUnit.SECONDS);
```
Comment (humkum) [inline store/src/main/java/org/apache/rocketmq/store/kv/CompactionStore.java:101]: Different compactionLog files are distinguished by topic and queueId, and do compaction in every compactionLog. Does this mean messages with the same keys must send to the same queue?

Comment (ltamber) [reply store/src/main/java/org/apache/rocketmq/store/kv/CompactionStore.java:101]: Yes.

File: store/src/main/java/org/apache/rocketmq/store/queue/SparseConsumeQueue.java [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+        final String subfolder) {
+        super(topic, queueId, storePath, mappedFileSize, defaultMessageStore, subfolder);
+    }
+
+    @Override
+    public void recover() {
+        MappedFile lastMappedFile = this.mappedFileQueue.getLastMappedFile();
+        if (lastMappedFile == null) {
+            return;
+        }
```
Comment (complone) [inline store/src/main/java/org/apache/rocketmq/store/queue/SparseConsumeQueue.java:56]: Do I need to add ```synchronized``` or ```lock``` to ensure thread safety when restoring a ByteBuffer from native?

Comment (ltamber) [reply store/src/main/java/org/apache/rocketmq/store/queue/SparseConsumeQueue.java:56]: the method `recover` invoked by only one thread, I think there is no need to add `synchronized` or `lock`

**Pull request merged**